refactor(eip7928): refactor net zero filtering in BALs

gurukamath opened the pull request:
## 🗒️ Description
This PR refactors the net zero refactoring mechanism in BALs and tightens the code up by introducing the following changes

1. Make assertions such that capturing pre-state only happens in tx level frame or higher.
2. Use all post-values from the relevant `*_changes` attributes from the frame (instead of looking into the state)
3. Remove the `capture_pre_nonce` function since this is not used anywhere
4. Standardize the filtering code across balance, code and storage in the `filter_net_zero_frame_changes` function.


## ✅ Checklist

- [x] All: Ran fast `tox` checks to avoid unnecessary CI fails, see also [Code Standards](https://eest.ethereum.org/main/getting_started/code_standards/) and [Enabling Pre-commit Checks](https://eest.ethereum.org/main/dev/precommit/):
    ```console
    uvx tox -e static
    ```
- [x] All: PR title adheres to the [repo standard](https://eest.ethereum.org/main/getting_started/contributing/?h=contri#commit-messages-issue-and-pr-titles) - it will be used as the squash commit message and should start `type(scope):`.

gurukamath committed:
refactor(eip7928): refactor net zero filtering in BALs

gurukamath commented on src/ethereum/forks/amsterdam/state_tracker.py:
```diff
@@ -521,44 +499,50 @@ def filter_net_zero_frame_changes(
     ----------
     tx_frame :
```
Since this can also be called on a block level frame (while processing withdrawals), we might need to rethink the `tx_frame` variable name and other related details.

fselmo reviewed (dismissed):
Nice cleanups. This lgtm!

I'm a bit hesitant spending too much more time on any more state changes here before the [larger state refactoring](https://github.com/ethereum/execution-specs/issues/1865#issuecomment-3642660766) changes are implemented since we will likely use that instead. What do you think is a good line to draw for us as far as cleaning things up here and waiting for the more general state refactor?

This is definitely an improvement though, let's get this in 👍🏼

gurukamath commented:
> What do you think is a good line to draw for us as far as cleaning things up here and waiting for the more general state refactor?

I don't think we should be actively looking to make any design changes right now. This particular change just crossed my mind as I was looking at the specs in order to write a test for a different scenario. And figured it was a quick win that tightens our implementation. (possibly eliminating any subtle bugs)

I wouldn't mind having minor adjustments to the spec implementation from time to time (especially to fix logical inconsistencies) but agree with you that most of our focus right now should be test implementation and releases.

fselmo commented:
> I wouldn't mind having minor adjustments to the spec implementation from time to time (especially to fix logical inconsistencies) but agree with you that most of our focus right now should be test implementation and releases.

Yeah, agreed 👍🏼. sgtm 😃

fselmo commented:
@gurukamath I found some failing tests with this refactor, specifically `test_selfdestruct_created_in_same_tx_with_revert`:

```
FAILED tests/cancun/eip6780_selfdestruct/test_selfdestruct_revert.py::test_selfdestruct_created_in_same_tx_with_revert[fork_Amsterdam-blockchain_test_from_state_test-outer_selfdestruct_before_inner_call-same_tx] - execution_testing.test_types.block_access_list.exceptions.BlockAccessListValidationError: Expected balance_changes to be empty but found [BalBalanceChange(rlp_override=None, block_access_index=1, post_balance=1)]
```

I haven't had a chance to dig into the issue yet but we should address this.

gurukamath committed:
refactor(eip7928): handle selfdestruct correctly

gurukamath commented:
> I haven't had a chance to dig into the issue yet but we should address this.

The issue here was that self-destruct was not being handled properly. We need to delete the balance changes (just like we do with nonce and code changes) for self-destructed accounts. I have updated the PR to make necessary changes.

fselmo commented:
> The issue here was that self-destruct was not being handled properly. We need to delete the balance changes (just like we do with nonce and code changes) for self-destructed accounts. I have updated the PR to make necessary changes.

It looks like now `test_bal_self_destruct` is failing. I also have a PR with a lot more boundary cases for selfdestruct. I expect some would fail there as well. PR #1954 also includes a fix for selfdestruct that I introduced in my [refactor PR](https://github.com/ethereum/execution-specs/pull/1897) (I moved the `track_address` into only the COLD path by accident... fixed [here](https://github.com/ethereum/execution-specs/pull/1954/changes#diff-28a1b575fd7c3d82832c0826cf58a881101643543d35c123c78ca65202152c23L643-R643)), but it looks like this failure is unrelated to that - even if I apply the one line fix here it will still fail `test_bal_self_destruct`.

gurukamath commented:
> It looks like now `test_bal_self_destruct` is failing. I also have a PR with a lot more boundary cases for selfdestruct. I expect some would fail there as well. PR #1954 also includes a fix for selfdestruct that I introduced in my [refactor PR](https://github.com/ethereum/execution-specs/pull/1897) (I moved the `track_address` into only the COLD path by accident... fixed [here](https://github.com/ethereum/execution-specs/pull/1954/changes#diff-28a1b575fd7c3d82832c0826cf58a881101643543d35c123c78ca65202152c23L643-R643)), but it looks like this failure is unrelated to that - even if I apply the one line fix here it will still fail `test_bal_self_destruct`.

Ah!! The curious case of the pre-funded account, created and self-destructed in the same tx. Don't know how I missed that. Have updated the PR.

Also, I ran this re-factor on top of what you have in #1954 and all the tests there are passing.

fselmo reviewed (approved):
Much cleaner, ty! 🚀
